### PR TITLE
Rework checksums

### PR DESCRIPTION
**Pull request by sbcgua**
The concept:
- separate checksums from repo metadata -> DBType `REPO_CS`
- `|` separated text file instead of XML. Why `|` - I was planning to use just space, but there are abap objects with space in a standard system. Also with `:`, and with `#` and with tab (!!!) and also with `|` - but just one ... well, I had to stop somewhere to still keep it readable
- decouple checksums as much as possible
- also fix some subtle bugs

# update 2022-03-26

Conceptually I tried not to change the functionality. So from user perspective all remain the same, this is more a refactoring. I noticed some potential bugs on the way but that can be solved later. Also the CS class is now unit-testable which suggests a better process to fix bugs.

- potentially solves #5154, due to the fact that checksums are not a prt of repo meta anymore. Though needs a real test (this item was my personal motivator for the change, hehe ;P ). See `chunk_lib->render_repo_palette`
- `zif_abapgit_repo_checksums` and `zcl_abapgit_repo_checksums` respectively incapsulates all the checksum related stuff. So the repo class is free from checksums logic (except factory method `checksums()`).
  - the class contains local classes `lcl_checksum_serializer` for the serializeing logic and `lcl_update_calculator` for the local/remote comparison logic. The comparison logic is the same as before though a bit destructured into a couple of methods
  - zcl_repo - minus 193 code lines, yey !
  - all covered with UTs (though not extremely intensively, leaves room to improve)
- `zif_abapgit_persist_repo_cs` repserens the db logic, implemented in `zcl_abapgit_persistence_repo` (why not?)
  - `zcl_abapgit_persist_factory` can create `get_repo_cs`, though in high level code it is supposed to use `zif_repo->checksums( )`
- checksums data is stored in `ZABAPGIT` table under `REPO_CS` type. The structure is:

```
#repo_name#abapgit
@
/|.abapgit.xml|7c0506a2af34fd1b42027e0288198a00d933d3d4
CLAS|ZCL_ABAPGIT_ADT_LINK|$ABAPGIT_UTILS
/src/utils/|zcl_abapgit_adt_link.clas.abap|909437692c00ea4b93c708d2572eb4f014756b2e
/src/utils/|zcl_abapgit_adt_link.clas.xml|b171192bcffce8241f08dbd70ed2d5bcfab17f76
CLAS|ZCL_ABAPGIT_AJSON|$ABAPGIT_JSON
/src/json/|zcl_abapgit_ajson.clas.abap|f2806341d814073d58046b1388fc9005665c5085
/src/json/|zcl_abapgit_ajson.clas.locals_imp.abap|b9eb49aabf187f1ed3a27b09f3da69199480ab83
/src/json/|zcl_abapgit_ajson.clas.testclasses.abap|1b9df4a8526fcb01bc31cc818e3256e75e155292

```
- Key spec:
  - `#key#value` is metadata. Currently added repo name meta purely for debug and possibility to see the first line in the DB_VIEW page to identify the CS without cross searching the metadata. Potentially metadata can be used in future for something else too
  - line starting from `/` are files
  - grouped by lines not started from `/` - OBJ_TYPE/OBJ_NAME/PACKAGE tripple, exception is `@` - root "object" without object attached (for `.abapgit` and Ko)
  - separator is `|`, appears there are objects in SAP containing spaces, tabs (!!), and some other nice looking symbols ... well ...
- DB page shows checksums properly

2022-03-26_13h46_55
2022-03-26_13h46_39

Further plans and doubts:
- caching: now CS is always loaded from DB on demand. Maybe it is OK. Or maybe no. No idea, time will show.
- next step is removing the `list_favorites` from repo_srv. But separately as it might initiate discussions
- **Migration !** I think there is no need for specific migration. Errors in checksums are currently ignored. So on next pull/push the checksums will be just recalculated. This is IMHO. Though of course it may lead to some minor confusion on the way (all `MM` statuses). Maybe this area can be improved, ideas welcomed.

### CODE REVIEW

**Comment by larshp** [comment]
rework? which parts does it rework?

separate checksums from repo metadata only?

or does it aim to get checksums working? If fixing bugs, we need to decide what exactly the checksums are, and what they are for. I dont think they contain or should be used to determine the local vs remote status, initially it was only meant for determining if overwriting local changes

**Comment by larshp** [comment]
ref #3997 #4555

**Comment by sbcgua** [comment]
ref #5154 also, so adresses low performance on loading long list of large repos

> separate checksums from repo metadata only?

Mainly this one.

And will simplify further checksums discussions and fixing

**Comment by sbcgua** [comment]
I've been thinking on this a couple of weeks already in background, so maybe let me clarify what I'm going to do from the beginning:
1) So the trigger was #5154 - the repo list has "Fav only" by default and F2 does not work properly (does not show all the repos) in this mode. Which is VERY inconvenient
2) Also there was a mention (#4555) that the performance is slow on large number of repos with large number of objects, because checksums is a part of meta data
3) This PR adresses these 2 problems
4) As the result the checksums will be more readable = more easier to identify issues (I already saw a couple)
5) and unit testable
6) Which will unlock further improvements

Architechture: in progress, but the main idea is to separate checksum login in a separate persistence layer and separate class, loosely or not-at-all coupled with the repo. Current idea is to have something like this:
```
lo_repo->checksums()->get()
lo_repo->checksums()->rebuild()
lo_repo->checksums()->update()
```
But it may change on the way

**Comment by larshp** [comment]
cool 👍 I guess "separate persistence layer" is still the good old ZABAPGIT db table?

**Comment by sbcgua** [comment]
Yes, that code is ready actually:
https://github.com/abapGit/abapGit/pull/5328/files#diff-5d9ea8c5a9e863c8f3cc2fea3fe705da29591476644110021ca5142c3033fcacR312
https://github.com/abapGit/abapGit/pull/5328/files#diff-459b532af877eebe8e6e9091dfa911349b868de0cfa7df368452e92aff5b6d94R1

**Comment by sbcgua** [comment]
> How will we get from the existing metadata to the new repo_cs entries?

Good question.

I think migration is not needed. CSs are not critical for proper execution. They will self-fix on the next commit/pull. Maybe an info banner should be added to indicate that "CSs are missing, will be regenerated" or smth like this.

By the way for the same reason I don't raise exceptions in the persistence (this code is published already)

**Comment by flaiker** [comment]
> forcing all interface methods implementation might be unnesessary in for testing classes. It is even not required by abap itself (there was some docs about it even). Multiple empty methods just clutter the code. Also not insisting: both approaches have + and -

If I recall correctly: It is not required for global interfaces (you get a syntax warning). I would guess this is for SAP to be able to add new methods to BAdIs without breaking customer code. Local interfaces must be fully implemented (syntax error) unless  the implementing class is `FOR TESTING` and the addition `PARTIALLY IMPLEMENTED` is used which is >= 740SP02.

Since the test classes are removed from the merged build I guess it would still work but you'll also get lots of syntax warnings in normal abapGit development that you cannot turn off. Did you not get these?

image

**Comment by sbcgua** [comment]
Yes, I get the warnings and this is one of the minuses :( Though no dumps. I wonder what's worse: these warnings or extra clutter in the code ... both are not that good :)

And thanks for the input !

**Comment by sbcgua** [comment]
hmmm .... :) Maybe it just means that the interface should be splitted into several ones ?

image

File: src/repo/zcl_abapgit_repo_checksums.clas.testclasses.abap [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  METHOD zif_abapgit_repo~get_files_remote.
+    rt_files = mt_remote_files.
+  ENDMETHOD.
+
+ENDCLASS.
+
+CLASS ltcl_local_file_builder DEFINITION FINAL.
+  PUBLIC SECTION.
+    DATA mt_tab TYPE zif_abapgit_definitions=>ty_files_item_tt.
+    METHODS add IMPORTING iv_str TYPE STRING.
```
[the remark this answers is not in the document]
Comment (larshp) [reply src/repo/zcl_abapgit_repo_checksums.clas.testclasses.abap:169]: bad bot

**Comment by larshp** [comment]
abaplint, paramters => https://github.com/abaplint/abaplint/issues/1400

**Comment by flaiker** [comment]
> I wonder what's worse: these warnings or extra clutter in the code ... both are not that good :)

I often do Advanced -> Syntax Check before committing so I'll always see those warnings there regardless of which code I touched in the repo. The clutter you only see when looking at that specific test class include.

> Maybe it just means that the interface should be splitted into several ones ?

:+1: for trying to keep interfaces small, though in this case it might be very difficult. Another idea would be to create a global test double you can inherit from locally so it doesn't fill up the test class include with the empty methods.

**Comment by sbcgua** [comment]
> Another idea would be to create a global test double you can inherit from locally ...

Yeah, that's what I do in other my projects. Poor man's "partially implemented" :)) @larshp what do you think ?

> I often do Advanced -> Syntax Check

I guess you're right, this is a more frequent case so should have higher priority. :+1:

**Comment by mbtools** [comment]
we should not have any warnings. otherwise, we need to track which warnings are ok and which need to be resolved. 

there are other test classes with the same problem (many empty interface methods). global test doubles would simplify those too, but abapmerge would have to exclude them.

**Comment by sbcgua** [comment]
@larshp why unit tests fail ? I excluded the class from transpilation https://github.com/abapGit/abapGit/pull/5328/files#diff-07d62516798d61561e897825f0c27cb28834955e2b02c0618b07a418349344e7

File: test/abap_transpile.json [hunk 2]
```diff
@@ -249,7 +249,11 @@
       {"object": "ZCL_ABAPGIT_SERVICES_BASIS", "class": "ltcl_create_package", "method": "raise_error_if_package_exists", "note": "Void type: SCOMPKDTLN"},
       {"object": "ZCL_ABAPGIT_SERVICES_BASIS", "class": "ltcl_create_package", "method": "package_given_in_popup", "note": "Void type: SCOMPKDTLN"},
       {"object": "ZCL_ABAPGIT_SERVICES_BASIS", "class": "ltcl_create_package", "method": "package_not_created_when_canc", "note": "Void type: SCOMPKDTLN"},
-      {"object": "ZCL_ABAPGIT_SERVICES_BASIS", "class": "ltcl_create_package", "method": "package_created_when_confirm", "note": "Void type: SCOMPKDTLN"}
+      {"object": "ZCL_ABAPGIT_SERVICES_BASIS", "class": "ltcl_create_package", "method": "package_created_when_confirm", "note": "Void type: SCOMPKDTLN"},
+
+      {"object": "ZCL_ABAPGIT_REPO_CHECKSUMS", "class": "ltcl_test_checksum_serializer"},
```
[the remark this answers is not in the document]
Comment (sbcgua) [reply test/abap_transpile.json:254]: Ah ... I though it is kind of mask. But that mean that I have to add methods each time I add a new test ... inconvenient. Do you mind if I adjust the transpiler to access whole class also ? (maybe you could point where is it in it's code ?)

[the remark this answers is not in the document]
Comment (larshp) [reply test/abap_transpile.json:254]: https://github.com/abaplint/transpiler/blob/main/packages/transpiler/src/unit_test.ts#L84

[the remark this answers is not in the document]
Comment (larshp) [reply test/abap_transpile.json:254]: first two unit tests running, 
image

File: test/abap_transpile.json [hunk 3]
```diff
@@ -249,7 +249,11 @@
...
+      {"object": "ZCL_ABAPGIT_REPO_CHECKSUMS", "class": "ltcl_test_checksums"},
+      {"object": "ZCL_ABAPGIT_REPO_CHECKSUMS", "class": "ltcl_update_calculator_test"}
```
Comment (larshp) [inline test/abap_transpile.json:256]:
```suggestion
      {"object": "ZCL_ABAPGIT_REPO_CHECKSUMS", "class": "ltcl_test_checksums", "method": "rebuild_w_dot_abapgit"},
      {"object": "ZCL_ABAPGIT_REPO_CHECKSUMS", "class": "ltcl_test_checksums", "method": "rebuild_simple"}
```

Comment (larshp) [reply test/abap_transpile.json:256]: after https://github.com/abapGit/abapGit/pull/5351 is merged, just these two needs to be skipped

Comment (larshp) [reply test/abap_transpile.json:257]: I think this suggestion is still relevant

**Comment by mbtools** [comment]
What's missing? Anything I can help with?

**Comment by sbcgua** [comment]
I will finish, but later ... due to obvious reasons ...

**Comment by sbcgua** [comment]
OK, so I think it is ready for review. Not all the ideas were implemented yet. But the PR is already huge and the remaining changes are just a bit opinionated, so I decided to separate them (e.g. I want to remove `list_favorites` from `repo_srv`). To guide through the PR I added an updated comment in the main PR message above.

review and opinions are very appreciated

**Comment by sbcgua** [comment]
... keep the update branch right on me for the moment because of some specifics of my current dev flow (local gitea + local dev serv) ... will unblock (and rebase) later, when the PR is more or less approved

**Comment by mbtools** [comment]
Good progress! I will give it a try the next days.

- Format: Maybe add a "record indicator" column `0|meta...`, `1|CLAS...`, `2|/src/...`. This allows for future extensions like `3|branch...`
- Caching: It's a select single on `zabapgit` and parsing the data in one loop. Even with thousands of objects this should be fast. > No cache
- Migration: If I understand it right, the checksum data does not change. A one time mass migration would be better. Without it there will be a long time with intermediate states (i.e. forever ;-) ).

**Comment by sbcgua** [comment]
Format: hmm, I'd like to keep it as simple as possible. In order to have a group for further extension I'd prefer to understand what kind of extension this is, not to overengineer. E.g. branches can be done with meta sections. `#key#value` format gives a log of flexibility e.g. additional structure for specific `key`. While levels might make certain confusion if eventually higher logical level would have "lower" digits

Caching: yeah agree. Had the same thought in mind.

Migration: hmm, partially agree. Let's wait for some more opinions. I'm just a bit lazy to do this :) Although it is not hugely difficult.

**Comment by sbcgua** [comment]
Btw do you see migration as actual read old data, convert, write new ? Or just force "recalculate" checksums for all repos ?

**Comment by mbtools** [comment]
read, convert, write

**Comment by sbcgua** [comment]
A bit offtopic, but: why `zif_abapgit_persist_repo~list` reads all repos (mo_db->list_by_type( zcl_abapgit_persistence_db=>c_type_repo )) instead of reading individual repo record by key ? Seems inefficient on large number of repos ... @larshp

**Comment by sbcgua** [comment]
OK, the migration is also ready. But I suggest reviewing it separately. As it is kind of one way operation. https://github.com/sbcgua/abapGit/pull/56
Otherwise all who will try to test this PR will loose their checksums. @mbtools @larshp

**Comment by sbcgua** [comment]
The migration is a bit hacky ... e.g. I had to add `zcl_abapgit_persistence_repo->rewrite_repo_meta`. But I think it might be useful in case of other meta data changes

**Comment by mbtools** [comment]
backup of db-table before testing 💯 
https://docs.abapgit.org/ref-database-util.html#backup--restore

**Comment by sbcgua** [comment]
Heh, I didn't know this feature. Too late for me ... :)
Well, let me know if to merge the migration code into this PR

**Comment by sbcgua** [comment]
merged migration code in here. Though beware with DB changes and use the hint above: https://github.com/abapGit/abapGit/pull/5328#issuecomment-1082447792

**Comment by sbcgua** [comment]
```
Error: > kernel_cx_assert [Error]: [object Object]
    at Function.assert_equals (file:///home/runner/work/abapGit/abapGit/output/cl_abap_unit_assert.clas.mjs:194:20)
    at ltcl_test_checksums.rebuild_simple 
```

**Comment by larshp** [comment]
add to skip, https://github.com/sbcgua/abapGit/blob/rework_checksums/test/abap_transpile.json#L103

**Comment by sbcgua** [comment]
did anyone have a chance to check this code ?

**Comment by mbtools** [comment]
not yet. waiting for my queue to empty so I can do a clean pull of this PR.

**Comment by larshp** [comment]
yea, sorry, I'll have a look tomorrow/this week

**Review by larshp — APPROVED** [review]

**Comment by larshp** [comment]
approved, I'll merge this PR tomorrow

**Comment by larshp** [comment]
for the conversion, I'm running ZABAPGIT with this code, and ZABAPGIT_STANDALONE with an old version, the old version adds old style checksums.

I'd expect that when running the new version, the old style checksums are removed, even though they have been converted again. This might not be a big issue, it will fix itself over time.

@sbcgua any thoughts?

**Comment by sbcgua** [comment]
Yes, you're right, this is designed like that. The criteria for conversion is not presence of checksums in XML but presence of REPO_CS key in table. https://github.com/abapGit/abapGit/pull/5328/files#diff-e95fbb70e7f41933ac27ce120512519a9a3bd1a25eb00383167b2f10a1aad958R80

But this is intentional. Otherwise the migration will always unconditionally read and parse all repos which will create a startup delay. This is kind of done anyway ... on repo list page ... but then it would be 2 times ... so ... I decided to do it indirectly. I think it is a good compromise. Maybe it makes sense to metion in the changelog that recommendation is to renew the standalone version as well with this release. But anyway this is a rare case.

Option is to double check the presense of checksums somewhere in repo server code ... but that would distribute the migration code ... ugh ... imho no :)

**Comment by larshp** [comment]
cool, merging
